backend/browseruserinput.go

```diff
@@ -0,0 +1,132 @@
+// message.
+
+// Message types
+const ChooseCenterCardMsgType string = "choose-center-card"
+const ChoosePlayerMsgType string = "choose-player"
+const ChoosePlayerInsteadOfCenterMsgType string = "choose-player-instead-of-center"
+
+type BrowserUserInput struct {
+	// lock protects the expecting variable. So that writes to the
+	// variable are read atomically
+	lock sync.RWMutex
+
+	// message takes a message from goroutine for the user request and
+	// gives it to the game engine
+	message chan interface{}
+
+	// expecting is the type of message the game engine is expecting from this
+	// player right now, or "" if the game engine is not expecting a
+	// message from this player
+	expecting string
+}
+
+func NewBrowserUserInput() *BrowserUserInput {
+	return &BrowserUserInput{
+		message: make(chan interface{}, 0),
+	}
+}
+
+func (input *BrowserUserInput) ChooseCenterCard(string) int {
+	input.lock.Lock()
+	input.expecting = ChooseCenterCardMsgType
+	input.lock.Unlock()
+
+	card := <-input.message
+	input.expecting = ""
+	return card.(int)
+}
+
+func (input *BrowserUserInput) ChoosePlayer(string, []string) string {
+	input.lock.Lock()
+	input.expecting = ChoosePlayerMsgType
+	input.lock.Unlock()
+
+	player := <-input.message
+	input.expecting = ""
+	return player.(string)
+}
+
+func (input *BrowserUserInput) DoesChoosePlayerInsteadOfCenter(string) bool {
+	input.lock.Lock()
+	input.expecting = ChoosePlayerInsteadOfCenterMsgType
+	input.lock.Unlock()
+
+	choice := <-input.message
+	input.expecting = ""
+	return choice.(bool)
+}
+
+func (input *BrowserUserInput) Prompt() string {
+	input.lock.RLock()
+	defer input.lock.RUnlock()
+	return input.expecting
+}
+
+func (input *BrowserUserInput) ReceiveMessage(msgType string, msgBody interface{}) error {
+	input.lock.RLock()
+	defer input.lock.RUnlock()
+
+	_, msgIsInt := msgBody.(int)
+	_, msgIsBool := msgBody.(bool)
+	_, msgIsString := msgBody.(string)
+
+	if input.expecting == "" {
+		return errors.New(fmt.Sprintf("Game engine is not expecting input " +
+			"from this player"))
+	} else if msgType != input.expecting {
+		return errors.New(fmt.Sprintf(
+			"Received message of type \"%s\". Expected message of "+
+				"type \"%s\"", msgType, input.expecting))
+	} else if msgType == ChooseCenterCardMsgType && !msgIsInt {
+		return errors.New(fmt.Sprintf(
+			"Message body for ChooseCenterCardMsgType should be <int>. "+
+				"Received %x", msgBody))
+	} else if msgType == ChoosePlayerMsgType && !msgIsString {
+		return errors.New(fmt.Sprintf(
+			"Message body for ChoosePlayerMsgType should be <string>. "+
+				"Received %x", msgBody))
+	} else if msgType == ChoosePlayerInsteadOfCenterMsgType && !msgIsBool {
+		return errors.New(fmt.Sprintf(
+			"Message body for ChoosePlayerInsteadOfCenterMsgType "+
+				" should be <bool>. Received %x", msgBody))
+	} else {
+		return errors.New(fmt.Sprintf("Unknown message type \"%s\"", msgType))
+	}
+	select {
+	case input.message <- msgBody:
+		break
```

Comment:
why not return here?

Reply:
In other languages I like reducing the number of times i return from a function. I find that having multiple return points confuses the control flow of the function. Go actively supports multiple returns though, so i can return early here.